blueflag/dataparcels
# Release/crazy constrictor

Addresses #196  #204 #215 #205 #213

Hooks are finally here!

## dataparcels
- **BREAKING CHANGE** Validation
  - `Validation` renamed to `validation` to match all other function exports
  - now exports a ParcelValueUpdater function rather than `{modifyBeforeUpdate: Function}`
  - now sets top level `meta.invalidList`
  - now doesnt validate until first attempted submit like most forms do
  - now blocks submit when used with `useParcelForm` and `useParcelBuffer`
- Boundary splits are temporarily disabled, see https://github.com/blueflag/dataparcels/pull/219/commits/dfcbd8998bb77dfd92aaa4c96bb91dd80ff3b548
- add `Parcel.onChangeDOMCheckbox()` #196 
- add `Parcel.spreadDOMCheckbox()` #196 
- use `Object.is()` in `ChangeRequest.hasValueChanged()` so `NaN`s are accurately tested for equality

## react-dataparcels
- add `useParcelState` hook (similar to ParcelHoc)
- add `useParcelBuffer` hook (similar to ParcelBoundaryHoc)
- add `useParcelForm` hook (new hook to provide pre-build form-like behaviours)
-  **BREAKING CHANGE** ParcelBoundary is now made out of `useParcelBuffer`
  - `debugBuffer` is gone
  - `debugParcel` is gone
  - `onCancel` is gone
  - `onRelease` is gone
  - `modifyBeforeUpdate` is deprecated, superseded by `beforeChange`, now accepts non-array functions
  - 2nd argument of childRenderer now gives a `ParcelBufferControl`, not a `ParcelBoundaryControl`
- `ParcelHoc` and `ParcelBoundaryHoc` are deprecated
  - You're better off making an ad-hoc hoc out of the hooks you want

## dataparcels-docs
- Use `dcme-gatsby` #213
- Refine a lot of pages #205 
- Add new API pages for hooks
